configure_pll() might be in a rush to test PLL_LOCK

Around line 293 of src/drivers/sifive_fe310-g000_pll.c:

```
    /* Wait for PLL to lock */
    while((__METAL_ACCESS_ONCE(pllcfg) & PLL_LOCK) == 0) ;
```

Section 6.5 of the FE310-G002 Manual 1p4 says:

> The PLL provides a lock signal which is set when the PLL has achieved lock, and which can be
> read from the most-significant bit of the pllcfg register. The PLL requires up to 100 μs to
> regain lock once enabled, and the lock signal will not necessarily be stable during this initial lock
> period so should only be interrogated after this period.

So this code seems to be a little aggressive, not waiting 100uS before entering the wait loop. I have not confirmed why yet, but a test program that configures the PLL to 320MHz infrequently hangs apparently when setting the PLL. Hardware is SparkFun Red-V with R8 corrected to 100 ohms (power supply filter for PLLVDD).

I appreciate a 100uS delay might be hard to arrange here in Metal given it seems to depend on other hardware configuration (LFROSC?) or CPU clock rate.

I'm realizing I'm not sure how to use this branch with the Windows Eclipse workspace that I'm using. I downloaded the branch, and used Project > Properties > Freedom Studio > SDK... to change the METAL_SOURCE_PATH to the top of the branch dir. Eclipse attempts to build but becomes unhappy with 17 errors like below. Any pointers to docs or wisdom on how to do this?

In file included from C:/Users/Dana/Downloads/freedom-metal-pll-lock-race/freedom-metal-pll-lock-race/src/drivers/riscv_plic0.c:12:
C:/Users/Dana/Downloads/freedom-metal-pll-lock-race/freedom-metal-pll-lock-race/src/drivers/riscv_plic0.c: In function '__metal_plic0_claim_interrupt':
C:/Users/Dana/Downloads/freedom-metal-pll-lock-race/freedom-metal-pll-lock-race/src/drivers/riscv_plic0.c:22:43: error: 'METAL_RISCV_PLIC0_CONTEXT_BASE' undeclared (first use in this function); did you mean 'METAL_RISCV_PLIC0_PENDING_BASE'?
         (__metal_io_u32 *)(control_base + METAL_RISCV_PLIC0_CONTEXT_BASE +
                                           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
C:/Users/Dana/Downloads/freedom-metal-pll-lock-race/freedom-metal-pll-lock-race/metal/io.h:8:47: note: in definition of macro '__METAL_ACCESS_ONCE'
 #define __METAL_ACCESS_ONCE(x) (*(__typeof__(*x) volatile *)(x))

--- a/metal/drivers/riscv_clint0.h
+++ b/metal/drivers/riscv_clint0.h
@@ -21,4 +21,7 @@ struct __metal_driver_riscv_clint0 {
 };
 #undef __METAL_MACHINE_MACROS
 
+int __metal_driver_riscv_clint0_command_request(
+    struct metal_interrupt *controller, int command, void *data);
+
 #endif

--- a/metal/drivers/sifive_clic0.h
+++ b/metal/drivers/sifive_clic0.h
@@ -42,4 +42,7 @@ struct __metal_driver_sifive_clic0 {
 };
 #undef __METAL_MACHINE_MACROS
 
+int __metal_driver_sifive_clic0_command_request(
+    struct metal_interrupt *controller, int command, void *data);
+
 #endif

--- a/src/drivers/sifive_fe310-g000_pll.c
+++ b/src/drivers/sifive_fe310-g000_pll.c
@@ -10,6 +10,7 @@
 
 #include <metal/init.h>
 #include <metal/machine.h>
+#include <metal/machine/platform.h>
 
 #include <metal/drivers/sifive_fe310-g000_pll.h>
 #include <stdlib.h>
@@ -270,6 +271,19 @@ static int find_closest_config(long ref_hz, long rate) {
     return closest_index;
 }
 
+/* The PLL needs 100 usec to stabilize before we test PLL_LOCK. Since LFROSC
+ * on all targets with the FE310-G000 PLL runs at 32768 Hz, we need to wait
+ * at least
+ *
+ *   ceil(100 usec * 32768 ticks/sec * 1 sec / 1000000 usec) = 4 ticks
+ *
+ * of mtime before we test PLL_LOCK.
+ *
+ * TODO: Determine the mtime timebase at compile or runtime and use that
+ * here.
+ */
+#define PLL_LOCK_WAIT_TICKS 4
+
 /* Configure the PLL and wait for it to lock */
 static void configure_pll(__metal_io_u32 *pllcfg, __metal_io_u32 *plloutdiv,
                           const struct pll_config_t *config) {
@@ -296,6 +310,30 @@ static void configure_pll(__metal_io_u32 *pllcfg, __metal_io_u32 *plloutdiv,
     }
 
     __METAL_ACCESS_ONCE(pllcfg) &= ~(PLL_BYPASS);
+
+    /* Wait for the PLL to stabilize before testing it for lock */
+#ifdef __METAL_DT_RISCV_CLINT0_HANDLE
+    unsigned long long mtime, mtime_end;
+    __metal_driver_riscv_clint0_command_request(__METAL_DT_RISCV_CLINT0_HANDLE,
+                                                METAL_TIMER_MTIME_GET, &mtime);
+    mtime_end = mtime + PLL_LOCK_WAIT_TICKS;
+    while (mtime <= mtime_end) {
+        __metal_driver_riscv_clint0_command_request(
+            __METAL_DT_RISCV_CLINT0_HANDLE, METAL_TIMER_MTIME_GET, &mtime);
+    }
+#elif __METAL_DT_RISCV_CLIC0_HANDLE
+    unsigned long long mtime, mtime_end;
+    __metal_driver_sifive_clic0_command_request(__METAL_DT_RISCV_CLIC0_HANDLE,
+                                                METAL_TIMER_MTIME_GET, &mtime);
+    mtime_end = mtime + PLL_LOCK_WAIT_TICKS;
+    while (mtime <= mtime_end) {
+        __metal_driver_sifive_clic0_command_request(
+            __METAL_DT_RISCV_CLIC0_HANDLE, METAL_TIMER_MTIME_GET, &mtime);
+    }
+#else
+#pragma message(                                                               \
+    No handle for CLINT or CLIC found, PLL might race with lock signal !)
+#endif
 
     /* Wait for PLL to lock */
     while ((__METAL_ACCESS_ONCE(pllcfg) & PLL_LOCK) == 0)

